src/main/java/org/apache/sling/rewriter/impl/TransformerFactoryServiceTracker.java

```diff
@@ -58,8 +62,11 @@ public TransformerFactoryServiceTracker(final BundleContext bc, final String ser
     @Override
     public Object addingService(ServiceReference reference) {
         final boolean isGlobal = isGlobal(reference);
+        LOGGER.debug("Adding service {}, isGlobal={}", reference, isGlobal);
         if ( isGlobal ) {
-            this.cacheIsValid = false;
+            synchronized (this) {
```

Comment:
I think just synchronizing the change of the flag is not enough; this needs to be synchronized with the state change (registering the service reference in the tracker). The same for removal.
I have the feeling that it is easier to rewrite this outdated code completely

Reply:
Thanks Carsten for pointing out the problem.. AFAIU, you meant that updates should be in sync with the super.addingService/removedService? I have made some changes. Please let me know if you were pointing something else or there is something else which needs to be improved.